handbook/engineering/observability/monitoring_pillars.md

```diff
@@ -93,23 +98,21 @@ Consider for example an admin who finds on a "Search" dashboard that "search que
 
 If instead this information is communicated only on a "Frontend" dashboard, it becomes immediately clear they can check the frontend logs first, which is where they will find next steps.
 
-#### What you should do instead
+#### What you should do
 
 Consider "when the metric I am monitoring is bad, where do I want the site admin to look?" Usually, this is just the service that emits the metric itself.
 
 In other words, you should add the metric you are monitoring to an existing service's dashboard.
 
 This also allows our [high-level alerting metrics](https://docs.sourcegraph.com/admin/observability/metrics_guide), which are generated from the same observables that each dashboard is generated from, to describe the health of Sourcegraph per-service (instead of per-service + topic like "HTTP").
 
-### FAQ: I have a purely information metric, I don't want to define an alert for it. How do I do that?
-
-This violates the second pillar of monitoring at Sourcegraph:
+### Graphs should have associated alerts
 
-> Adding a graph/panel that visualizes a metric, without defining an associated alert, is forbidden.
+Adding a graph/panel that visualizes a metric, without defining an associated alert, is discouraged. This includes purely informational graphs.
```

Comment:
I don't agree with this. Lots of informational graphs don't need to ping anyone with an alert, and other graphs will be new enough that we don't know what good or bad is yet.

Reply:
> Lots of informational graphs don't need to ping anyone with an alert

Yep! There are certainly exceptions, could use your help on this! https://github.com/sourcegraph/about/pull/2000/files#r523020140

> I don't agree with this.

Do you mind elaborating on this more in the context of the existing "Explanation" section? @slimsag wrote a lot of this and I trust he had a good reason for all this, and I was personally pretty sold by this explanation :)

Reply:
> By saying "I do not want to define an alert" what we are actually doing is shying away from the question of "how should someone interpret what this number means?" and instead saying "I will keep this information private, just for myself"

Being able to interpreting the graph doesn't mean you want to alert on it. And when you do want to alert on it, it can take time actually monitoring the service in production to learn what useful thresholds are for alerts. Best guesses are often as good as no guess.

Reply:
There are also plenty of cases that give humans useful signal as to how a system is behaving, but are too variable or inconsistent to reliably alert on. for example i definitely want to be able to see at a glance the number of healthy  requests to a given service, but i don't think you could define an alert on that metric that works for all use cases. I think separating graphs from informational vs actionable is worth thinking about, and there's a lot you can gain from informational graphs or dashboards when debugging alerts from actionable ones.

Reply:
> it can take time actually monitoring the service in production to learn what useful thresholds are for alerts. Best guesses are often as good as no guess.

I"m not sure I follow - doesn't performing a query for a metric without a dashboard give you a look at a how the metric behaves in production already? I think the the expectation might have been to have a service export a metric in production for a while, _then_ create a dashboard based on how it behaves to ship, by which point you might have a semi-informed guess.

> Being able to interpreting the graph doesn't mean you want to alert on it

That's true too - I think there's room for improvement here overall, @pecigonzalo and I bounced the idea of properly supporting "informational" metrics with an alternative to `PossibleSolutions` (something like `Interpreting`?) to generate docs for purely information graphs. One idea was we could expand the guideline here to include something like "either you describe how to interpret it through docs, or through alerts" - how does that sound?

Reply:
>  I think the expectation might have been to have a service export a metric in production for a while, then create a dashboard based on how it behaves to ship, by which point you might have a semi-informed guess. 

That's not the workflow I'm used to and that I think is most productive. Metrics and dashboards evolve hand in hand. I want to iterate on a dashboard, over time, in production. Perhaps there's a way to separate the "published" and "high level summary" part of a dashboard, to be consumed mostly by site admins, from the rest?

> "either you describe how to interpret it through docs

Grafana has a description field for a panel, can't we use that?

Reply:
> That's not the workflow I'm used to and that I think is most productive. Metrics and dashboards evolve hand in hand. I want to iterate on a dashboard, over time, in production. Perhaps there's a way to separate the "published" and "high level summary" part of a dashboard, to be consumed mostly by site admins, from the rest?

We have to keep in mind this is not just an internal dashboard and are part of the overall product which we ship to customers monthly and remain there for a period of time. I agree that dashboards evolve over time, but I believe its better to think about them as we think about our services, we iterate over in production by committing, PR and shipping. Dashboards are part of the site-admin experience and we should treat them, like our APIs and site-config, etc.
If we have the tooling to support it, do you think developing locally against the remote Prometheus works? If not, what are the pitfalls?

> Grafana has a description field for a panel, can't we use that?

I would prefer to have it in our documentation instead, as we are lacking a lot information on that front, many metrics I dont know how to interpret and It would be helpful to have some information.

Reply:
> If we have the tooling to support it, do you think developing locally against the remote Prometheus works? If not, what are the pitfalls?

Not as well, since we can't share the dashboard with others as easily. Your concerns would be better addressed by having "internal" groups in each dashboard that don't have to abide by the same standards.

> I would prefer to have it in our documentation instead, as we are lacking a lot information on that front, many metrics I dont know how to interpret and It would be helpful to have some information.

I think both make sense.

Reply:
> Not as well, since we can't share the dashboard with others as easily.

How would you normally share it? Right now you should be able to share the dashboard JSON generated by the generator

> I would prefer to have it in our documentation instead, as we are lacking a lot information on that front, many metrics I dont know how to interpret and It would be helpful to have some information.

Link to docs in description, similar to those generated for alerts? :)

> Perhaps there's a way to separate the "published" and "high level summary" part of a dashboard, to be consumed mostly by site admins

You can do so already by creating a new group in the generator (even better now without required alerts)

Reply:
> Not as well, since we can't share the dashboard with others as easily.

How would you normally share it? Right now you should be able to share the dashboard JSON generated by the generator.

Reply:
In response to this feedback, and [the other thread](https://github.com/sourcegraph/about/pull/2000#discussion_r523050126), I have completely removed this pillar in favour of adding more emphasis on "just use the generator" to ensure documentation is in place: 8662cba269a6255dfc4a16fa7f9242dcb8d25a1a

This will require some implementation work to enforce docs on graphs without alerts. I have added it to the PR description, and will open up more specific issues when we've finalized this PR :)

Reply:
> How would you normally share it? Right now you should be able to share the dashboard JSON generated by the generator.

Link to the live dashboard. Example: https://sourcegraph.com/-/debug/grafana/d/QvLV8J2Gk/git-server-cloud?orgId=1&from=1605082470899&to=1605083727190&var-shard=All

Reply:
> Not as well, since we can't share the dashboard with others as easily. Your concerns would be better addressed by having "internal" groups in each dashboard that don't have to abide by the same standards.

Could you elaborate on this? I'm not mapping how internal dashboards address my comment regarding develop in production and the workflow used for metrics at other places in which the dashboards are not shipped to customers and released as part of the product.
For example, we dont develop it in production by just changing the live files, we instead send a PR which can be reviewed, etc. I think these dashboards have to be changed in the same way.

I do think having some internal dashboards that we can iterate over a bit faster, likely not tied to alerts but rather information that we can use to debug .com and customer environments. Or just experimental dashboards that are not "ready" yet as discussed in the other thread and as we talked about in our conversation when we said "Is there room for 2 kinds of dashboards?".

Coming back to the development experience, is the main point the inability to share the dashboard directly via a link? How are we currently doing when developing other interfaces?

Reply:
> Coming back to the development experience, is the main point the inability to share the dashboard directly via a link? How are we currently doing when developing other interfaces?

Yeah, quick iteration and visual feedback, and sharing via a link with other teammates for more feedback, analysis, etc.

Reply:
> > Coming back to the development experience, is the main point the inability to share the dashboard directly via a link? How are we currently doing when developing other interfaces?
> 
> Yeah, quick iteration and visual feedback, and sharing via a link with other teammates for more feedback, analysis, etc.

I imagine for other interfaces it's just screenshots at the moment, right? Assuming that work is just as iterative, I'm not sure if it is really that uncomfortable to go without the link sharing. This seems like a reasonable workflow:

1. Port-forward prometheus
2. `git checkout ...`
3. `./dev/grafana.sh`
4. marvel at dashboards

It's not necessary to spin up everything `dev/start`, so dashboards arguably already have the best sharing/collaboration experience at the moment (maybe?)

> For example, we dont develop it in production by just changing the live files, we instead send a PR which can be reviewed, etc. I think these dashboards have to be changed in the same way.

+1, and this seems like a good reason to _not_ use the Grafana editor to make edits in the Grafana database and share a link with teammates (I think this is the only way to get the workflow Tomas envisions)

Reply:
> I do think it will be good to have some internal dashboards that we can iterate over a bit faster, likely not tied to alerts but rather information that we can use to debug .com and customer environments. Or just experimental dashboards that are not "ready" yet as discussed in the other thread and as we talked about in our conversation when we said "Is there room for 2 kinds of dashboards?".

Do we envision needing experimental dashboards for customers? I don't think they upgrade frequently enough for this to be a good way to iterate - if we do need this, we can send them dashboard JSON as required, and for dot-com do experimental dashboard provisioning via ConfigMap: https://github.com/sourcegraph/about/pull/2000#discussion_r523864481 

This workflow would then look like this:

* Provision experimental dashboards in via `deploy-sourcegraph-dot-com`. Iterate freely without need for review because we don't ship these to customers.
* Iterate on experimental dashboards in customer instances by sending them dashboard JSON. Iterate freely because we can just keep sending them more dashboard JSON without having to ask them to upgrade to a new, shipped Grafana image.
* Remove the current custom dashboards shipped with the Grafana image to drive eventually "promoting" experimental dashboards to generator dashboards, and to have more impetus to improve the generator to enable this if required (to avoid becoming increasingly dependent on raw dashboards)

Reply:
SGTM

Reply:
Created action items to make this workflow happen in the PR description - resolving this conversation for now :) Thanks for the discussion!